---
title: "SIAMCAT input files formats"
author:
-   name: "Konrad Zych, Jakob Wirbel, and Georg Zeller"
    affiliation: "EMBL Heidelberg"
    email: "georg.zeller@embl.de"
date: "2018-05-16"
output: BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{SIAMCAT.holdout}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{ASCII}
---

# Introduction

This vignette illustrates how to read in and input your own data to the
`SIAMCAT` package. We will cover reading in text files from the disk, formatting
them and creating an object of `siamcat-class` of them. 

An object of `siamcat-class` is the centerpiece of the package. All of the input
data and result are stored inside. The structure of the object is described in
the [siamcat-class object](#siamcat-class-object) section.


# Loading your data into R

## SIAMCAT format files

There are three input files in SIAMCAT format:

### Features file

The features file is a tsv file that is organized as follows: 
features (in rows) x samples (in columns).

First row should contain sample names, while the first column should contain 
feature labels (e.g. taxonomic identifiers). The remaining entries are expected 
to be real values >= 0 that quantify the abundance of each feature in each 
sample:

| | Sample_1 | Sample_2  | Sample_3 | Sample_4  | Sample_5 |
| --- | ---:| ---:| ---:| ---:| ---:|
| **Feature_1** | 0.59 | 0.71 | 0.78 | 0.61 | 0.66 |
| **Feature_2** | 0.00 | 0.02 | 0.00 | 0.00 | 0.00 |
| **Feature_3** | 0.02 | 0.00 | 0.00 | 0.00 | 0.00 |
| **Feature_4** | 0.34 | 0.00 | 0.43 | 0.00 | 0.00 |
| **Feature_5** | 0.56 | 0.56 | 0.00 | 0.00 | 0.00 |

An example of such a file is atached to the SIAMCAT package:

```{r message=FALSE}
library(SIAMCAT)
fn.in.feat  <- system.file(
    "extdata",
    "feat_crc_zeller_msb_mocat_specI.tsv",
    package = "SIAMCAT"
)
```

We can access the file with a dedicated SIAMCAT function `read.features`:
```{r results="hide", warning=FALSE}
feat  <- read.features(fn.in.feat)
```

### Metadata file

The metadata file a tsv file organized as follows:
samples (in rows) x metadata (in columns):

First row should contain metadata variables names, while the first column should
contain sample names.

| | Age | Gender | BMI |
| --- | ---:| ---:| ---:|
| **Sample_1** | 52 | 1 | 20|
| **Sample_2** | 37 | 1 | 18 |
| **Sample_3** | 66 | 2 | 24 |
| **Sample_4** | 54 | 2 | 26 | 
| **Sample_5** | 65 | 2 | 30 | 

An example of such a file is atached to the SIAMCAT package:
```{r message=FALSE}
fn.in.meta  <- system.file(
    "extdata",
    "num_metadata_crc_zeller_msb_mocat_specI.tsv",
    package = "SIAMCAT"
)
```

We can access the file with a dedicated SIAMCAT function `read.meta`:
```{r results="hide", warning=FALSE}
meta  <- read.meta(fn.in.meta)
```

### Label file
The label file is a tsv file with labels and converts it into a label object.

First row is expected to be #BINARY:1=[label for cases]; 
-1=[label for controls]. Second row should contain the sample identifiers as 
tab-separated list (consistent with feature and metadata).

Third row is expected to contain the actual class labels (tab-separated): 
1 for each case and -1 for each control:

| #BINARY:1=cancer;-1=healthy |
| :---| :---| :---| :---| :---|
| Sample_1 | Sample_2  | Sample_3 | Sample_4  | Sample_5 |
| 1 | 1 | -1 | -1 | 1 |

An example of such a file is atached to the SIAMCAT package:
```{r message=FALSE}
fn.in.label <- system.file(
    "extdata",
    "label_crc_zeller_msb_mocat_specI.tsv",
    package = "SIAMCAT"
)
```

We can access the files with the dedicated SIAMCAT functions and directly
construct a SIAMCAT object containing the microbial features, the patient`s
labels, and metadata for the patients.
```{r results="hide", warning=FALSE}
label <- read.labels(fn.in.label)
```

The label can also be created from one of the metadata columns. For example, to
create a label from a `crc_stage` column, treating value `0` as a case and the
other values as a control:
```{r results="hide", warning=FALSE}
label <- create.label.from.metadata(meta, "crc_stage", case = 0)
```

### Creating a siamcat-class object

Out of the features, label and (optionally) metadata objectst we can create a 
`siamcat-class`. The structure of the object is described inthe 
[siamcat-class object](#siamcat-class-object) section.
```{r results="hide", warning=FALSE}
siamcat <- siamcat(feat, label, meta)
```


## LEfSe format files

[LEfSe](https://bitbucket.org/biobakery/biobakery/wiki/lefse) is a tool for 
identification of associations between micriobial features and up to two 
metadata. LEfSe uses LDA (linear discriminant analysis).

LEfSe input file is a tsv file. First few rows contain the metadata. The 
following row contains sample names and further ones are occupied by features. 
The first column contains row names:

| label | healthy | healthy  | healthy | cancer  | cancer |
| --- | ---:| ---:| ---:| ---:| ---:|
| **age** | 52 | 37  | 66 | 54  | 65 |
| **gender** | 1 | 1  | 2 | 2  | 2 |
|**Sample_info** | Sample_1 | Sample_2  | Sample_3 | Sample_4  | Sample_5 |
| **Feature_1** | 0.59 | 0.71 | 0.78 | 0.61 | 0.66 |
| **Feature_2** | 0.00 | 0.02 | 0.00 | 0.00 | 0.00 |
| **Feature_3** | 0.02 | 0.00 | 0.00 | 0.00 | 0.00 |
| **Feature_4** | 0.34 | 0.00 | 0.43 | 0.00 | 0.00 |
| **Feature_5** | 0.56 | 0.56 | 0.00 | 0.00 | 0.00 |

An example of such a file is atached to the SIAMCAT package:
```{r message=FALSE}
fn.in.lefse<- system.file(
    "extdata",
    "LEfSe_crc_zeller_msb_mocat_specI.tsv",
    package = "SIAMCAT"
)
```

SIAMCAT has a dedicated function to read LEfSe format files. The `read.lefse` 
function will read in the input file and extract metadata and features data:
```{r results="hide", warning=FALSE}
metaAndFeat <- read.lefse(fn.in.lefse, rows.meta = 1:6, row.samples = 7)
meta <- metaAndFeat$meta
feat <- metaAndFeat$feat
```

We can then create a lable object from one of the columns of the meta object and
 create a `siamcat` object:
```{r results="hide", warning=FALSE}
label <- create.label.from.metadata(meta, "label", case = "cancer")
siamcat <- siamcat(feat, label, meta)
```

## MaAsLin format files

[MaAsLin](https://bitbucket.org/biobakery/biobakery/wiki/maaslin) is a tool for 
identification of associations between micriobial features and complex  
metadata. MaAsLin multivariate analysis.

MaAsLin can have two types of input files: 

a) two tsv files, one for metadata and one for features - those can be used in
 SIAMCAT just like with [SIAMCAT format files](#SIAMCAT-format-files) 
 `read.features` and `read.meta`;

b) a single PCL file that can be read like the 
[LEfSe format files](#LEfSe-format-files) with the `read.lefse` function. 

## metagenomeSeq format files
[metagenomeSeq](http://bioconductor.org/packages/metagenomeSeq/) is an R package
 to determine differentially abundant features between multiple samples. 
 
There are two ways to input data into metagenomeSeq:

a) two tsv files, one for metadata and one for features - those can be used in
 SIAMCAT just like with [SIAMCAT format files](#SIAMCAT-format-files) 
 `read.features` and `read.meta`. The features file needs to be transposed with 
`transpose = TRUE` parameter:
```{r results="hide", warning=FALSE, eval=FALSE}
fn.in.feat  <- system.file(
    "extdata",
    "CHK_NAME.otus.count.csv",
    package = "metagenomeSeq"
)
feat <- read.features(fn.in.feat, transpose = TRUE)
```

b) BIOM format file, that can be used in SIAMCAT as decribed in the 
[following section](BIOM-format-files)

## BIOM format files
The BIOM format files can be added to SIAMCAT via phyloseq. First the file 
should be imported using function `import_biom`. Then a phyloseq object should 
be imported to a siamcat object as descibed in the 
[Creating a siamcat object of a phyloseq object](#Creating-a-siamcat-object-of-a-phyloseq-object) section.

## Creating a siamcat object of a phyloseq object
The siamcat object extends a phyloseq object, so creating a siamcat object from 
a phyloseq object is really straightforward. This can be done with `siamcat` 
constructor function. First, however, we need to create a label object:
```{r results="hide", warning=FALSE, eval=TRUE}
data("GlobalPatterns") ## phyloseq example data
label <- create.label.from.metadata(sample_data(GlobalPatterns),
                                    column = "SampleType",
                                    case = "Feces")
```

And then we will run the constructor function:
```{r results="hide", warning=FALSE, eval=TRUE}
siamcat <- siamcat(GlobalPatterns,label)
```

# siamcat-class object
A siamcat class object is the centerpiece of the package where all the data is 
stored:

![internal make-up of a siamcat object](./siamcat.png)
Rectangle depict slots of the object while class of the object stored in the 
slot is given in an oval. There are three obligatory slots - phyloseq, label and
orig_feat - marked with thick borders. 

In order to explain the siamcat object better we will show how each of the slots
is filled.

## phyloseq, label and orig_feat slots
The phyloseq and label slots are obligatory. The phyloseq slot is an object of 
class phyloseq (that is described in the help file of the `phyloseq` class (can 
be accessed e.g. by typing into R console: `help('phyloseq-class')`). The label 
slot contains a list. This list has specific construction 
(`help('label-class')`). The `orig_feat` contains object of the `otu_table` 
class (`help('otu_table-class')`) with original feature table (that remains 
unmodified throught the run  of the package).

Thephyloseq, label and orig_feat are filled while siamcat object is first 
created with `siamcat` function. The minimum valid call of the `siamcat`
function requires a feature table (of `otu_table` class e.g. returned by 
`read.features` function) and a label object (e.g. returned by `read.label` 
function or created from metadata table with `create.label.from.metadata` 
function):
 ![](./allCr.png)

## All the other slots
Other slots are filled during the run of the siamcat workflow:
![](./Slots_create.png)

## Accessing and assigning slots

Each slot in siamcat can be accessed with a `slot_name(siamcat)` e.g `eval_data`
 slot with `eval_data(siamcat)`. There is one notable exception of `phyloseq` 
slot that due to technical reasons can be accessed with `physeq(siamcat)`.

Slots will be assigned by the SIAMCAT package functions. However, if for any 
reason a slot needs to be assigned outside of the worklow following construction
 can be used: `slot_name(siamcat) <- object_to_assign` e.g. to assing object 
`new_label` to `label` slot: `label(siamcat) <- new_label`.

## Slots inside the slots

There are two slots that have slots inside of them. First, `model_list` slot has
 `models` slot that contains the actual list of mlr models (and can be accessed
via `models(siamcat)`) and `model.type` - a character with the name of the 
method used to train the model (`model_type(siamcat)`).

The `phyloseq` slot has a complex structure. However, unless the `phyloseq` 
object is created outside of the SIAMCAT workflow, only two slots of phyloseq 
slot will be occupied: `otu_table` slot containing features table (that can be 
accessed with `features(siamcat)`) and `sam_data` slot containing metadata 
information (`meta(siamcat)`).

Additional slots inside the `phyloseq` slots do not have dedicated accessors, 
but can easily be reached once the `phyloseq` object is exported from the 
`siamcat` object:
```{r}
phyloseq <- physeq(siamcat)
tax_tab <- tax_table(phyloseq)
head(tax_tab)
```


# Session Info

```{r}
sessionInfo()
```
